### PR TITLE
Edit Tags - Ropes issue14

### LINKED ISSUE #14: Modify Tag List
#### As a user I want to edit the tag list for an image.

__Story__:

Image tags are awesome -- our user's love them! Tags help us make sense of the
plethora of images added to our system. But how annoying is it when our users
make spelling mistakes in their image tags? Super annoying! How awesome, and
useful would it be to allow for tags to be edited after an image is added?
Super awesome, and useful!

__Acceptance criteria__:
- [ ] There is a link to edit the tags for an image.
- [ ] The image whose tags are to be modified is displayed on the edit page.
- [ ] On the edit page, I can change the tags to any valid tag list.
- [ ] After saving the edit page with valid tags, my changes are visible
  throughout the rest of the application.

__Discussion__:
- [ ] Should an image's URL be editable?

__Dependencies__:
- Image Tags
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by gitlev**
Closes #14

### CODE REVIEW

File: app/controllers/images_controller.rb [hunk 1]
```diff
@@ -16,6 +16,8 @@ def index
     @images = @images.order(created_at: :desc)
   end
 
+  def show; end
```
Comment (wilsonyqm) [inline app/controllers/images_controller.rb]: this should be a separate commit

File: app/views/images/edit.html.erb [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<div class="main">
+  <div class="card image-card--large">
+    <%= image_tag(@image.url, class: 'image-card--large__img py-4') %>
+  </div>
+  <%= simple_form_for @image, html: {class: 'form form-horizontal'} do |f| %>
+    <%= f.input :tag_list, input_html: {value: @image.tag_list.to_s} %>
```
Comment (wilsonyqm) [inline app/views/images/edit.html.erb]: is `to_s` here necessary?

File: config/routes.rb [hunk 3]
```diff
@@ -1,7 +1,7 @@
 Rails.application.routes.draw do
   root 'images#index'
 
-  resources :images, only: [:index, :new, :create, :show, :destroy] do
+  resources :images, only: [:index, :new, :create, :show, :edit, :update, :destroy] do
```
Comment (wilsonyqm) [inline config/routes.rb]: you don't the `only` now

File: app/views/images/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<div class="main">
+  <div class="card image-card--large">
+    <%= image_tag(@image.url, class: 'image-card--large__img py-4') %>
```
Comment (wilsonyqm) [inline app/views/images/edit.html.erb]: is `py-4` class for mobile view?

File: app/views/images/show.html.erb [hunk 5]
```diff
@@ -6,8 +6,9 @@
         <%= render partial: 'tag-list', locals: { tags: @image.tag_list, style: 'tag-list--large' } %>
       </div>
       <div class="card-block image-card--large__btn-block">
-        <%= link_to 'Delete', image_path(@image), class: 'btn btn-danger image-card--large__btn image-card--large__btn--delete', 'data-method': 'delete', 'data-confirm': 'Are you sure you want to delete this image?', rel: 'nofollow' %>
-        <button type="button" class="btn btn-primary image-card--large__btn js-share-button" data-body-url="<%= modal_body_image_path %>" >Share</button>
+        <button type="button" class="btn btn-primary image-card--large__btn mx-1 js-share-button" data-body-url="<%= modal_body_image_path %>" >Share</button>
+        <%= link_to 'Edit', edit_image_path(@image), class: 'btn btn-success image-card--large__btn mx-1 js-edit-button' %>
```
Comment (wilsonyqm) [inline app/views/images/show.html.erb]: can you have one commit to change the order first, and then another commit to add `Edit`, the history will look better.

File: app/controllers/images_controller.rb [hunk 6]
```diff
@@ -1,5 +1,5 @@
 class ImagesController < ApplicationController
-  before_action :find_image_or_redirect, only: [:show, :send_email, :destroy]
+  before_action :find_image_or_redirect, only: [:show, :edit, :update, :send_email, :destroy]
```
Comment (wilsonyqm) [inline app/controllers/images_controller.rb]: it's confused the `send_email` is on both line 2 and 3

File: app/controllers/images_controller.rb [hunk 7]
```diff
@@ -29,6 +31,17 @@ def create
     render :new, status: :unprocessable_entity
   end
 
+  def edit; end
+
+  def update
+    if params[:image][:url].nil?
```
Comment (wilsonyqm) [inline app/controllers/images_controller.rb]: `nil?` or `present?`

File: app/controllers/images_controller.rb [hunk 8]
```diff
@@ -34,10 +34,10 @@ def create
   def edit; end
 
   def update
-    if @image.update_attributes(image_params)
-      flash.now[:success] = 'Success! The tags have been updated.'
-      render 'show'
+    if params[:image][:url].nil?
```
Comment (wilsonyqm) [inline app/controllers/images_controller.rb]: I think you don't need to check this, the strong params already consider this case.

File: app/views/images/edit.html.erb (hunk 2 above)
Comment (gitlev) [inline app/views/images/edit.html.erb]: yes, otherwise the list will be output without the comma delimiter

File: app/controllers/images_controller.rb (hunk 6 above)
Comment (gitlev) [inline app/controllers/images_controller.rb]: thank you I've removed it from the second 👍

File: config/routes.rb (hunk 3 above)
Comment (gitlev) [inline config/routes.rb]: 👍

File: app/views/images/edit.html.erb (hunk 4 above)
Comment (gitlev) [inline app/views/images/edit.html.erb]: nope py means padding on the y axis

File: app/controllers/images_controller.rb (hunk 8 above)
Comment (gitlev) [inline app/controllers/images_controller.rb]: 👍

File: app/views/images/show.html.erb (hunk 5 above)
Comment (gitlev) [inline app/views/images/show.html.erb]: 👍

File: app/controllers/images_controller.rb (hunk 1 above)
Comment (gitlev) [inline app/controllers/images_controller.rb]: 👍

File: test/models/image_test.rb [hunk 9]
```diff
@@ -1,8 +1,17 @@
...
+  def test_extra_whitespace_removed_from_tags
+    setup_webstub 200, 'image/jpg', 'http://www.example.com'
+    image = Image.new(url: @url, tag_list: "  \tbear\t,    honey \n  ")
+    # acts-as-taggable-on/lib/acts_as_taggable_on/tag_list.rb
+    # clean method is removing the whitespace before the valid? is called
+    assert image.valid?
+    assert_equal image.tag_list[0], 'bear'
+    assert_equal image.tag_list[1], 'honey'
+  end
+
```
Comment (wilsonyqm) [inline test/models/image_test.rb]: there should another invalid test

File: test/flows/images_crud_test.rb [hunk 10]
```diff
@@ -1,6 +1,37 @@
 require 'flow_test_helper'
 
 class ImagesCrudTest < FlowTestCase
+  test 'edit image tags' do
```
Comment (wilsonyqm) [inline test/flows/images_crud_test.rb]: should have an invalid test here too

File: test/models/image_test.rb (hunk 9 above)
Comment (gitlev) [inline test/models/image_test.rb]: the invalid test is on line 44 and called ```test_tag_list_empty```

Comment (wilsonyqm) [inline test/models/image_test.rb]: Ohh, nice
